### PR TITLE
Slash command support

### LINKED ISSUE #205: Implement slash command support
Discord has released new "slash commands" which allow us to integrate commands directly with the API: https://discord.com/developers/docs/interactions/slash-commands

This means:

- Implementing the new `INTERACTION_CREATE` gateway event
- Implementing the endpoints documented at https://discord.com/developers/docs/interactions/slash-commands#endpoints and associated data types at https://discord.com/developers/docs/interactions/slash-commands#data-models-and-types

### PR DESCRIPTION
**Pull request by jchristgit**
This pull request implements slash command support, by implementing the new gateway event and accompanying API routes.

Closes #205.

- `16:37:00.679 [warn]  UNHANDLED GATEWAY DISPATCH EVENT TYPE: APPLICATION_COMMAND_CREATE, %{application_id: 999, description: "ed, man! man, ed", guild_id: 999, id: 999, name: "edit"}` is currently undocumented. Therefore, nothing for that here.

### CODE REVIEW

**Review by Kraigie — COMMENTED** [review]
This looks good! I think we should have a section somewhere in the docs that describes how these slash commands work in a little more depth, possibly with some examples. Thoughts?

File: lib/nostrum/struct/application_command_interaction_data_option.ex [hunk 1]
```diff
@@ -0,0 +1,94 @@
+defmodule Nostrum.Struct.ApplicationCommandInteractionDataOption do
+  @moduledoc "Struct for command invocation arguments."
+
+  # seasons greetings from `AbstractBeanModelParserFactory.java`.
```
Comment (Kraigie) [inline lib/nostrum/struct/application_command_interaction_data_option.ex:4]: This is very well written! Consider a career in literature if the whole programming thing doesn't work out. 

I think we both know this should probably be removed, but honestly I don't hate it. Maybe we meet half way and just keep the first paragraph? 😸

Comment (jchristgit) [reply lib/nostrum/struct/application_command_interaction_data_option.ex:4]: I dropped it: 63eca89

**Comment by jchristgit** [comment]
I have added extra documentation in 74f8850bf8dc42ed0033ea12f019a15f475de8d9.

Any idea how we should deal with the credo warning? To me it feels sort of like a false positive, or at least one that we should ignore. Of course, I can also factor that out

**Comment by Kraigie** [comment]
> I have added extra documentation in [74f8850](https://github.com/Kraigie/nostrum/commit/74f8850bf8dc42ed0033ea12f019a15f475de8d9).
> 
> Any idea how we should deal with the credo warning? To me it feels sort of like a false positive, or at least one that we should ignore. Of course, I can also factor that out

Yeah I think we should probably add an ignore for that credo warning, it's definitely bogus.

File: lib/nostrum/struct/application_command_interaction_data_option.ex [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  # for the most part, these people are not used to people greeting them.
+  # greeting other people with a smile, saying good morning, saying good
+  # evening, all those little things are oil that keeps the cogs in our society
+  # running. people who are looking on the ground may just need that little bit
+  # of oil to get their cogs running smoothly again.
+  #
+  # speaking of oil, cars are very great. i had this idea recently where i made
+  # my car run erlang, with the premise that when i crash the car, it would
+  # just respawn at home. unfortunately, i do not believe that that is how that
+  # works.
```
Comment (sinni800) [inline lib/nostrum/struct/application_command_interaction_data_option.ex]: I love it.

Comment (jchristgit) [reply lib/nostrum/struct/application_command_interaction_data_option.ex]: That's fine sinni, thank you for your input

**Comment by jchristgit** [comment]
I increased credo mass threshold in https://github.com/Kraigie/nostrum/pull/207/commits/bfcfb9c1238e3592c2803b06da896f5c86594047

**Comment by Kraigie** [comment]
Thanks!

**Pull request merged**